jepsen/src/jepsen/control/k8s.clj

```diff
@@ -0,0 +1,111 @@
+(ns jepsen.control.k8s
+  "The recommended way is to use SSH to setup and teardown databases.
+  It's however sometimes conveniet to be able to setup and teardown
+  the databases using `kubectl` instead, which is what this namespace
+  helps you do. Use at your own risk, this is an unsupported way
+  of running Jepsen."
+  (:require [clojure.java.shell :refer [sh]]
+            [slingshot.slingshot :refer [throw+]]
+            [jepsen.control :as c]
+            [clojure.string :refer [split-lines, trim]]
+            [clojure.tools.logging :refer [info]])
+  (:import (jepsen.control Remote)))
+
+(defn exec
+  "Execute a shell command on a pod."
+  [context namespace pod-name {:keys [cmd] :as opts}]
+  (apply sh
+         "kubectl"
+         "exec"
+         (c/escape pod-name)
+         context
+         namespace
+         "--"
+         "sh"
+         "-c"
+         cmd
+         (if-let [in (:in opts)]
+           [:in in]
+           [])))
+
+(defn- path->pod
+  [pod-name path]
+  (str pod-name ":" path))
+
+(defn- unwrap-result
+  "Throws when shell returned with nonzero exit status."
+  [exc-type {:keys [exit] :as result}]
+  (if (zero? exit)
+    result
+    (throw+
+     (assoc result :type exc-type)
+     nil ; cause
+     "Command exited with non-zero status %d:\nSTDOUT:\n%s\n\nSTDERR:\n%s"
+     exit
+     (:out result)
+     (:err result))))
+
+(defn cp-to
+  "Copies files from the host to a pod filesystem."
+  [context namespace pod-name local-paths remote-path]
+  (doseq [local-path (flatten [local-paths])]
+    (->> (sh
+          "kubectl"
+          "cp"
+          (c/escape local-path)
+          (c/escape (path->pod pod-name remote-path))
+          context
+          namespace)
+         (unwrap-result ::copy-failed))))
+
+(defn cp-from
+  "Copies files from a pod filesystem to the host."
+  [context namespace pod-name remote-paths local-path]
+  (doseq [remote-path (flatten [remote-paths])]
+    (->> (sh
+          "kubectl"
+          "cp"
+          (c/escape (path->pod pod-name remote-path))
+          (c/escape local-path)
+          context
+          namespace)
+         (unwrap-result ::copy-failed))))
+
+(defn- or-parameter
+  "A helper function that encodes a parameter if present"
+  [p, v]
+  (if v (str "--" p "=" (c/escape v)) ""))
+
+(defrecord K8sRemote [context namespace]
+  Remote
+  (connect [this pod-name]
+    (assoc this
+           :context (or-parameter "context" context)
+           :namespace (or-parameter "namespace" namespace)
+           :pod-name pod-name))
+  (disconnect! [this]
+    (dissoc this :context :namespace :pod-name))
+  (execute! [this action]
+    (exec context namespace (:pod-name this) action))
+  (upload! [this local-paths remote-path _rest]
+    (cp-to context namespace (:pod-name this) local-paths remote-path))
+  (download! [this remote-paths local-path _rest]
+    (cp-from context namespace (:pod-name this) remote-paths local-path)))
+
+(def k8s
```

Comment:
I think it might be best to make this a function rather than a constant.

Reply:
Docker remote has it as def, so does ssh as well. Why would it be a function in this case?

Reply:
I missed that in the docker remote, but it shouldn't do it either. Trying to cut down on extraneous compile-time constants, and also leave ourselves escape hatches in case things need to be parameterized later.

Reply:
fixed